great_tables/data/__init__.py

```diff
@@ -1,1355 +1,1488 @@
-    "Cl_u_fac": "float64",
-    "Cl_A": "float64",
-    "Cl_B": "float64",
-    "Cl_n": "float64",
-    "Cl_t_low": "float64",
-    "Cl_t_high": "float64",
-}
-
-_photolysis_fname = DATA_MOD / "15-photolysis.csv"
-_photolysis_dtype = {
-    "cmpd_name": "object",
-    "cmpd_formula": "object",
-    "products": "object",
-    "type": "object",
-    "l": "float64",
-    "m": "float64",
-    "n": "float64",
-    "quantum_yield": "float64",
-    "wavelength_nm": "object",
-    "sigma_298_cm2": "object",
-}
-
-_nuclides_fname = DATA_MOD / "16-nuclides.csv"
-_nuclides_dtype = {
-    "nuclide": "object",
-    "z": "Int64",
-    "n": "Int64",
-    "element": "object",
-    "radius": "float64",
-    "radius_uncert": "float64",
-    "abundance": "float64",
-    "abundance_uncert": "float64",
-    "is_stable": "object",
-    "half_life": "float64",
-    "half_life_uncert": "float64",
-    "isospin": "object",
-    "decay_1": "object",
-    "decay_1_pct": "float64",
-    "decay_1_pct_uncert": "float64",
-    "decay_2": "object",
-    "decay_2_pct": "float64",
-    "decay_2_pct_uncert": "float64",
-    "decay_3": "object",
-    "decay_3_pct": "float64",
-    "decay_3_pct_uncert": "float64",
-    "magnetic_dipole": "float64",
-    "magnetic_dipole_uncert": "float64",
-    "electric_quadrupole": "float64",
-    "electric_quadrupole_uncert": "float64",
-    "atomic_mass": "float64",
-    "atomic_mass_uncert": "float64",
-    "mass_excess": "float64",
-    "mass_excess_uncert": "float64",
-}
+
+class _DataFrameConverter:
+    """ """
+
+    filepath: Path
+    dtype: dict[str, Any]
+    # `dtype` should contain Python types and rely on the hook `_convert_to_*()` to transform
+    # these types into the appropriate formats for different DataFrame implementations.
+
+    @staticmethod
+    def _convert_to_pandas(v: Any) -> str:
+        """
+        https://pandas.pydata.org/docs/user_guide/basics.html#dtypes
+        """
+        typs: dict[Any, str] = {int: "Int64", float: "Float64", str: "object"}
+        return typs.get(v, "object")
+
+    @classmethod
+    def to_pandas(cls):
+        import pandas as pd
+
+        dtype: dict[str, str] = {k: cls._convert_to_pandas(v) for k, v in cls.dtype.items()}
+        df = pd.read_csv(cls.filepath, dtype=dtype)
+        df.__doc__ = cls.__doc__
+        return df
+
+    @staticmethod
+    def _convert_to_polars(v: Any) -> Any:
+        import polars as pl
+
+        typs: dict[Any, Any] = {int: pl.Int64, float: pl.Float64, str: pl.String}
+        return typs.get(v, pl.String)
+
+    @classmethod
+    def to_polars(cls):
+        import polars as pl
+
+        schema: dict[str, Any] = {k: cls._convert_to_polars(v) for k, v in cls.dtype.items()}
+        df = pl.read_csv(cls.filepath, schema=schema)
+        df.__doc__ = cls.__doc__
+        return df
+
+    @staticmethod
+    def _convert_to_pyarrow(v: Any): ...
+
+    @classmethod
+    def to_pyarrow(cls): ...
+
+
+class Countrypops(_DataFrameConverter):
```

Comment:
Currently, this approach works by having the class _DataFrameConverter...

* specifying a filepath class attribute, etc..
* subclassing it to define those attributes

One risk of this approach is that a leans a bit hard on inheritance (see composition over inheritance). WDYT of an approach where things like Countrypops a instances of `_DataFrameConverter`, or something similar?

In this case you could add init arguments to `_DataFrameConverter`, like...

```python
class _DataFrameConverter:
    def __init__(self, filepath, dtype):
        self.filepath = filepath
        self.dtype = dtype

Countrypops = _DataFrameConverter(
    DATA_MOD / "01-countrypops.csv",
    dtype = {
        "country_name": str,
        "country_code_2": str,
        "country_code_3": str,
        "year": int,
        "population": int,
    }    
)
"""A docstring"""
```

Reply:
@machow , thank you for the excellent suggestion! Since we're moving towards favoring composition, renaming `_DataFrameConverter` to `_Dataset` seems more appropriate, and passing the docstrings as arguments aligns well with this approach. In addition, I've taken the liberty of implementing the `Pyarrow` conversion and added a `__repr__()` method, though I'm uncertain about its utility—feedback on this would be appreciated.

For the internal datasets, I recommend retaining the current parsing approach with `Pandas`. It appears to be more robust, and I've noticed that handling certain cases with `Polars`, such as `x-airquality.csv` and `x_locales.csv`, is a bit tricky. Since these datasets are used internally, we should have room to explore better solutions for these edge cases.